test/libyul/yulOptimizerTests/fullSuite/abi2.yul

```diff
@@ -0,0 +1,1218 @@
+				converted := convert_t_contract$_GnosisSafe_$710_to_t_uint160(value)
+			}
+		
+			function convert_t_contract$_GnosisSafe_$710_to_t_uint160(value) -> converted {
+				converted := cleanup_assert_t_uint160(value)
+			}
+		
+			function convert_t_contract$_Module_$1038_to_t_address(value) -> converted {
+				converted := convert_t_contract$_Module_$1038_to_t_uint160(value)
+			}
+		
+			function convert_t_contract$_Module_$1038_to_t_uint160(value) -> converted {
+				converted := cleanup_assert_t_uint160(value)
+			}
+		
+			function convert_t_enum$_Operation_$1949_to_t_uint8(value) -> converted {
+				converted := cleanup_assert_t_enum$_Operation_$1949(value)
+			}
+		
+				function copy_calldata_to_memory(src, dst, length) {
+					calldatacopy(dst, src, length)
+					// clear end
+					mstore(add(dst, length), 0)
+				}
+			
+				function copy_memory_to_memory(src, dst, length) {
+					let i := 0
+					for { } lt(i, length) { i := add(i, 32) }
+					{
+						mstore(add(dst, i), mload(add(src, i)))
+					}
+					if gt(i, length)
+					{
+						// clear end
+						mstore(add(dst, length), 0)
+					}
+				}
+			
+			function round_up_to_mul_of_32(value) -> result {
+				result := and(add(value, 31), not(31))
+			}
+			}
+// ----
+// fullSuite
+// {
+//     {
+//         let _1 := 1
+//         let _2 := mload(_1)
+//         let _3 := 0
+//         let _1017 := mload(_3)
+//         let abi_decode_value0_156 := _3
+//         if slt(sub(_2, _1017), 64)
+//         {
+//             revert(_3, _3)
+//         }
+//         {
+//             abi_decode_value0_156 := and(calldataload(_1017), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+//         }
+//         sstore(_3, abi_decode_value0_156)
+//         let _1019 := mload(8)
+//         let _1021 := mload(7)
+//         let abi_decode_value0_57 := _3
+//         let abi_decode_value1_58 := _3
+//         let abi_decode_value2_59 := _3
+//         let abi_decode_value3 := _3
+//         let abi_decode_value4 := _3
+//         if slt(sub(_1019, _1021), 128)
+//         {
+//             revert(_3, _3)
+//         }
+//         {
+//             abi_decode_value0_57 := and(calldataload(_1021), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+//         }
+//         {
+//             abi_decode_value1_58 := calldataload(add(_1021, 32))
+//         }
+//         {
+//             let abi_decode_offset_62 := calldataload(add(_1021, 64))
+//             if gt(abi_decode_offset_62, 0xffffffffffffffff)
+//             {
+//                 revert(_3, _3)
+//             }
+//             abi_decode_value2_59, abi_decode_value3 := abi_decode_t_bytes_calldata_ptr(add(_1021, abi_decode_offset_62), _1019)
+//         }
+//         {
+//             abi_decode_value4 := cleanup_revert_t_enum$_Operation_$1949(calldataload(add(_1021, 96)))
+//         }
+//         sstore(_1, abi_decode_value0_57)
+//         sstore(_1, abi_decode_value1_58)
```

Comment:
How hard would it be to remove redundant storage writes? At least in some settings it should be easy...

Reply:
The logic would be very similar to the RedundantAssignEliminator, but I hope we do not need it for the ABI coder. The only case where it should write twice to the same location is the free memory pointer and I guess that's not too bad.